### PR TITLE
Introduce SymbolicConstraint in SE

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
`Squash UseCSharp9` commit doesn't need a review

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs [hunk 1]
```diff
@@ -18,16 +18,48 @@
...
+using System.Linq;
+using System.Text;
+using SonarAnalyzer.Helpers;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn
 {
     public class SymbolicValue
     {
         private readonly int identifier;
+        private readonly Dictionary<Type, SymbolicConstraint> constraints = new();  // SymbolicValue can have only one constraint instance of specific type at a time
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs:32]: _FIXME (in this PR):_ Make it lazy

**Comment by pavel-mikula-sonarsource** [comment]
cc @costin-zaharia-sonarsource

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Constraints/LockConstraint.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.SymbolicExecution.Constraints
+{
+    public sealed class LockConstraint : SymbolicConstraint
+    {
+        public static readonly LockConstraint Held = new();
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Constraints/LockConstraint.cs:25]: Maybe wee need more than just held or released in order to be able to track if the reader or the writer lock is held/released?

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Constraints/LockConstraint.cs:25]: Held/Released should be enough for S2222. 
Tracking `Read/Write` kind of lock will be another (new) rule. See the google docs/FIXMEs/last point.

Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Constraints/LockConstraint.cs:25]: I missed that comment.

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**